pulsar-client-cpp/lib/MultiTopicsConsumerImpl.cc

```diff
@@ -0,0 +1,621 @@
+
+    if (messageListener_) {
+        listenerExecutor_->postWork(
+            boost::bind(&MultiTopicsConsumerImpl::internalListener, shared_from_this(), consumer));
+    }
+}
+
+void MultiTopicsConsumerImpl::internalListener(Consumer consumer) {
+    Message m;
+    messages_.pop(m);
+
+    try {
+        messageListener_(Consumer(shared_from_this()), m);
+    } catch (const std::exception& e) {
+        LOG_ERROR("Exception thrown from listener of Partitioned Consumer" << e.what());
+    }
+}
+
+Result MultiTopicsConsumerImpl::receive(Message& msg) {
+    Lock lock(mutex_);
+    if (state_ != Ready) {
+        lock.unlock();
+        return ResultAlreadyClosed;
+    }
+
+    if (messageListener_) {
+        lock.unlock();
+        LOG_ERROR("Can not receive when a listener has been set");
+        return ResultInvalidConfiguration;
+    }
+    messages_.pop(msg);
+    lock.unlock();
+
+    unAckedMessageTrackerPtr_->add(msg.getMessageId());
+    return ResultOk;
+}
+
+Result MultiTopicsConsumerImpl::receive(Message& msg, int timeout) {
+    Lock lock(mutex_);
+    if (state_ != Ready) {
+        lock.unlock();
+        return ResultAlreadyClosed;
+    }
+
+    if (messageListener_) {
+        lock.unlock();
+        LOG_ERROR("Can not receive when a listener has been set");
+        return ResultInvalidConfiguration;
+    }
+
+    if (messages_.pop(msg, milliseconds(timeout))) {
+        lock.unlock();
+        unAckedMessageTrackerPtr_->add(msg.getMessageId());
+        return ResultOk;
+    } else {
+        return ResultTimeout;
+    }
+}
+
+void MultiTopicsConsumerImpl::acknowledgeAsync(const MessageId& msgId, ResultCallback callback) {
+    if (state_ != Ready) {
+        callback(ResultAlreadyClosed);
+        return;
+    }
+    std::string& topicPartitionName = msgId.impl_->topicName_;
+    std::map<std::string, ConsumerImplPtr>::iterator iterator = consumers_.find(topicPartitionName);
+
+    if (consumers_.end() != iterator) {
+        unAckedMessageTrackerPtr_->remove(msgId);
+        iterator->second->acknowledgeAsync(msgId, callback);
+    } else {
+        LOG_ERROR("Message of topic: " << topicPartitionName << " not in unAckedMessageTracker");
+        callback(ResultUnknownError);
+        return;
+    }
+}
+
+void MultiTopicsConsumerImpl::acknowledgeCumulativeAsync(const MessageId& msgId, ResultCallback callback) {
+    callback(ResultOperationNotSupported);
+}
+
+MultiTopicsConsumerImpl::~MultiTopicsConsumerImpl() {}
+
+Future<Result, ConsumerImplBaseWeakPtr> MultiTopicsConsumerImpl::getConsumerCreatedFuture() {
+    return multiTopicsConsumerCreatedPromise_.getFuture();
+}
+const std::string& MultiTopicsConsumerImpl::getSubscriptionName() const { return subscriptionName_; }
+
+const std::string& MultiTopicsConsumerImpl::getTopic() const { return topic_; }
+
+const std::string& MultiTopicsConsumerImpl::getName() const { return consumerStr_; }
+
+void MultiTopicsConsumerImpl::setState(const MultiTopicsConsumerState state) {
```

Comment:
this should be a checkAndSet. Callers should have to specify which state they expect to be in before changing the state.

Reply:
thanks, some of the state could have "check and set", while like in close(), we could have multiple state choice before set.